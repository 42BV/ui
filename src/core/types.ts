/* istanbul ignore file */

export type Color =
  | 'primary'
  | 'secondary'
  | 'success'
  | 'info'
  | 'warning'
  | 'danger'
  | 'link'
  | 'muted'
  | 'dark'
  | 'light';

export type ModalSize = 'sm' | 'md' | 'lg' | 'xl';

type Side = 'top' | 'right' | 'bottom' | 'left';
type StartOrEnd = 'start' | 'end';
export type TippyPlacement = Side | `${Side}-${StartOrEnd}`;
