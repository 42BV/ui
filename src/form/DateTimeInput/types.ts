// See https://momentjs.com/docs/#/displaying/format/ for formatting options.
export type DateFormat = string | false; // When false it should not be used
export type TimeFormat = string | false; // When false it should not be used
