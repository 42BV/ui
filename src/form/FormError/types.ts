export type OnChange = (hasErrors: boolean) => void;