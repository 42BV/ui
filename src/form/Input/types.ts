export type Position = 'left' | 'right';
export type Mask = (string | RegExp)[];
