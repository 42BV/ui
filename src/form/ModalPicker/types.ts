export type AddButtonCallback<T> = () => Promise<T>;

export interface AddButtonOptions<T> {
  label: string;
  onClick: AddButtonCallback<T>;
}
