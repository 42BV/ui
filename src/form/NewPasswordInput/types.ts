export type NewPasswordInputRule =
  | 'lowercase'
  | 'uppercase'
  | 'number'
  | 'specialChar'
  | 'minimumLength'
  | 'noSpace';
