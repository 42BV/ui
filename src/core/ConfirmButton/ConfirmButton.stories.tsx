import React from 'react';
import { action } from '@storybook/addon-actions';

import { ConfirmButton } from './ConfirmButton';

export default {
  title: 'core/buttons/ConfirmButton',

  parameters: {
    component: ConfirmButton
  }
};

const AsButtonStory = () => {
  return (
    <div className="text-center">
      <p>When not in progress:</p>
      <ConfirmButton
        onConfirm={action('Accept clicked')}
        dialogText={
          <p>
            Are you sure you want to <strong>delete</strong> the user?
          </p>
        }
      >
        Delete user
      </ConfirmButton>

      <hr />

      <p>When in progress:</p>
      <ConfirmButton
        onConfirm={action('Accept clicked')}
        inProgress={true}
        dialogText={
          <p>
            Are you sure you want to <strong>delete</strong> the user?
          </p>
        }
      >
        Delete user
      </ConfirmButton>
    </div>
  );
};

export const AsButton = {
  render: AsButtonStory,
  name: 'as button'
};

const AsIconStory = () => {
  return (
    <div className="text-center">
      <p>When not in progress:</p>
      <ConfirmButton
        onConfirm={action('Accept clicked')}
        icon="delete"
        dialogText={
          <p>
            Are you sure you want to <strong>delete</strong> the user?
          </p>
        }
      />

      <hr />

      <p>When in progress:</p>
      <ConfirmButton
        onConfirm={action('Confirm clicked')}
        icon="delete"
        inProgress={true}
        dialogText={
          <p>
            Are you sure you want to <strong>delete</strong> the user?
          </p>
        }
      />
    </div>
  );
};

export const AsIcon = {
  render: AsIconStory,
  name: 'as icon'
};

const AsButtonWithIconStory = () => {
  return (
    <div className="text-center">
      <p>When not in progress:</p>
      <ConfirmButton
        onConfirm={action('Confirm clicked')}
        icon="delete"
        dialogText={
          <p>
            Are you sure you want to <strong>delete</strong> the user?
          </p>
        }
      >
        Delete user
      </ConfirmButton>

      <hr />

      <p>When in progress:</p>
      <ConfirmButton
        onConfirm={action('Confirm clicked')}
        icon="delete"
        inProgress={true}
        dialogText={
          <p>
            Are you sure you want to <strong>delete</strong> the user?
          </p>
        }
      >
        Delete user
      </ConfirmButton>
    </div>
  );
};

export const AsButtonWithIcon = {
  render: AsButtonWithIconStory,
  name: 'as button with icon'
};

const CustomTextAndColorStory = () => {
  return (
    <div className="text-center">
      <p>When not in progress:</p>
      <ConfirmButton
        color="info"
        onConfirm={action('Confirm clicked')}
        icon="delete"
        dialogText={
          <p>
            Are you sure you want to <strong>delete</strong> the user?
          </p>
        }
        text={{
          confirm: 'YES',
          cancel: 'NO',
          modalHeader: 'PLEASE SAY YES'
        }}
      >
        Delete user
      </ConfirmButton>

      <hr />

      <p>When in progress:</p>
      <ConfirmButton
        color="info"
        onConfirm={action('Confirm clicked')}
        icon="delete"
        inProgress={true}
        dialogText={
          <p>
            Are you sure you want to <strong>delete</strong> the user?
          </p>
        }
        text={{
          confirm: 'YES',
          cancel: 'NO',
          modalHeader: 'PLEASE SAY YES'
        }}
      >
        Delete user
      </ConfirmButton>
    </div>
  );
};

export const CustomTextAndColor = {
  render: CustomTextAndColorStory,
  name: 'custom text and color'
};

const FullWidthStory = () => {
  return (
    <div className="text-center">
      <p>When not in progress:</p>
      <ConfirmButton
        fullWidth={true}
        onConfirm={action('Accept clicked')}
        dialogText={
          <p>
            Are you sure you want to <strong>delete</strong> the user?
          </p>
        }
      >
        Delete user
      </ConfirmButton>

      <hr />

      <p>When in progress:</p>
      <ConfirmButton
        fullWidth={true}
        onConfirm={action('Accept clicked')}
        inProgress={true}
        dialogText={
          <p>
            Are you sure you want to <strong>delete</strong> the user?
          </p>
        }
      >
        Delete user
      </ConfirmButton>
    </div>
  );
};

export const FullWidth = {
  render: FullWidthStory,
  name: 'full width'
};
