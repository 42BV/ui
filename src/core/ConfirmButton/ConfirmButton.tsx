import React, { useState } from 'react';

import Button, { Props as ButtonProps } from '../Button/Button';
import ConfirmModal from '../ConfirmModal/ConfirmModal';

type Text = {
  /**
   * The text to show in the header, defaults to "Confirmation"
   *
   * @default Confirmation
   */
  modalHeader?: string;

  /**
   * The text to show as the cancel button's text, defaults to "Cancel"
   *
   * @default Cancel
   */
  cancel?: string;

  /**
   * The text to show as the ok button's text, defaults to "OK"
   *
   * @default OK
   */
  confirm?: string;
};

type Props = ButtonProps & {
  /**
   * The text you want to render inside of the dialog.
   */
  dialogText: React.ReactNode;

  /**
   * Callback which is triggered after the user has 'confirmed' that
   * the action should occur.
   *
   * Basically replaces the logic you would normally put in an `onClick`
   * event in a normal button.
   */
  onConfirm: () => void;

  /**
   * Optionally customized text to use within the component.
   */
  text?: Text;
};

/**
 * The ConfirmButton asks the user if he / she is sure he wants to
 * perform the action.
 *
 * The obvious use case is a delete button which asks the user if
 * he is sure he wants to delete something.
 */
export default function ConfirmButton({
  icon,
  children,
  fullWidth,
  size,
  dialogText,
  color = 'danger',
  onConfirm,
  inProgress,
  text = {},
  className
}: Props) {
  const [isOpen, setOpen] = useState(false);
  const { modalHeader, confirm, cancel } = text;

  function openModal(event: React.MouseEvent<HTMLElement>) {
    event.stopPropagation();
    setOpen(true);
  }

  function saveModal() {
    setOpen(false);
    onConfirm();
  }

  const buttonProps: ButtonProps = {
    onClick: openModal,
    color,
    inProgress: !!inProgress,
    icon,
    children,
    fullWidth,
    size
  };

  return (
    <div
      className={className}
      style={{
        display: buttonProps.icon && !buttonProps.children ? 'inline' : 'block'
      }}
    >
      <Button {...buttonProps} />

      <ConfirmModal
        isOpen={isOpen}
        onClose={() => setOpen(false)}
        onSave={() => saveModal()}
        label={modalHeader}
        modalText={dialogText}
        cancelText={cancel}
        confirmText={confirm}
      />
    </div>
  );
}
