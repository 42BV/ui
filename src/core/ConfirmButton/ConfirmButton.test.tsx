import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { ConfirmButton } from './ConfirmButton';
import { IconType } from '../types';

describe('Component: ConfirmButton', () => {
  describe('ui', () => {
    test('only button', () => {
      const { container } = render(
        <ConfirmButton
          onConfirm={() => undefined}
          dialogText={
            <p>
              Are you sure you want to <strong>delete</strong> the user?
            </p>
          }
        >
          Delete user
        </ConfirmButton>
      );

      expect(container).toMatchSnapshot();
    });

    test('only icon', () => {
      const { container } = render(
        <ConfirmButton
          onConfirm={() => undefined}
          icon="delete"
          dialogText={
            <p>
              Are you sure you want to <strong>delete</strong> the user?
            </p>
          }
        />
      );

      expect(container).toMatchSnapshot();
    });

    test('only icon in progress', () => {
      const { container } = render(
        <ConfirmButton
          onConfirm={() => undefined}
          inProgress={true}
          icon="delete"
          dialogText={
            <p>
              Are you sure you want to <strong>delete</strong> the user?
            </p>
          }
        />
      );

      expect(container).toMatchSnapshot();
    });

    test('button and icon', () => {
      const { container } = render(
        <ConfirmButton
          onConfirm={() => undefined}
          icon="delete"
          dialogText={
            <p>
              Are you sure you want to <strong>delete</strong> the user?
            </p>
          }
        >
          Delete user
        </ConfirmButton>
      );

      expect(container).toMatchSnapshot();
    });

    test('custom class, color and texts', () => {
      const { container } = render(
        <ConfirmButton
          className="bigger-better-and-uncut"
          color="info"
          onConfirm={() => undefined}
          icon="delete"
          dialogText={
            <p>
              Are you sure you want to <strong>delete</strong> the user?
            </p>
          }
          text={{
            confirm: 'YES',
            cancel: 'NO',
            header: 'PLEASE SAY YES'
          }}
        >
          Delete user
        </ConfirmButton>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('events', () => {
    type Props =
      | { icon: IconType; button?: never }
      | { button: string; icon?: never };

    function setup({ button, icon }: Props) {
      const onConfirmSpy = jest.fn();

      const props = {
        onConfirm: onConfirmSpy,
        dialogText: 'Are you sure you want a ConfirmButton?'
      };

      if (icon && button) {
        const { container } = render(
          <ConfirmButton icon={icon} {...props}>
            {button}
          </ConfirmButton>
        );
        return { container, onConfirmSpy };
      } else if (icon) {
        const { container } = render(<ConfirmButton icon={icon} {...props} />);
        return { container, onConfirmSpy };
      } else {
        const { container } = render(
          <ConfirmButton {...props}>{button}</ConfirmButton>
        );
        return { container, onConfirmSpy };
      }
    }

    it('should open the Modal when the ConfirmButton button is clicked', () => {
      setup({ button: 'Delete' });

      expect(
        screen.queryByText('Are you sure you want a ConfirmButton?')
      ).not.toBeInTheDocument();

      fireEvent.click(screen.getByText('Delete'));

      expect(
        screen.queryByText('Are you sure you want a ConfirmButton?')
      ).toBeInTheDocument();
    });

    it('should close the Modal when the Cancel button is clicked', () => {
      const { onConfirmSpy } = setup({ button: 'Delete' });

      fireEvent.click(screen.getByText('Delete'));

      expect(
        screen.queryByText('Are you sure you want a ConfirmButton?')
      ).toBeInTheDocument();

      fireEvent.click(screen.getByText('Cancel'));

      expect(
        screen.queryByText('Are you sure you want a ConfirmButton?')
      ).not.toBeInTheDocument();
      expect(onConfirmSpy).toHaveBeenCalledTimes(0);
    });

    it('should close and save the Modal when the Confirm button is clicked', () => {
      const { onConfirmSpy } = setup({ button: 'Delete' });

      fireEvent.click(screen.getByText('Delete'));

      expect(
        screen.queryByText('Are you sure you want a ConfirmButton?')
      ).toBeInTheDocument();

      fireEvent.click(screen.getByText('Confirm'));

      expect(
        screen.queryByText('Are you sure you want a ConfirmButton?')
      ).not.toBeInTheDocument();
      expect(onConfirmSpy).toHaveBeenCalledTimes(1);
    });
  });
});
