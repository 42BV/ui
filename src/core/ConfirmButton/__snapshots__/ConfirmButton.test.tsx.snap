// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ConfirmButton ui button and icon: Component: ConfirmButton => ui => button and icon 1`] = `
<div
  style={
    Object {
      "display": "block",
    }
  }
>
  <Button
    color="danger"
    icon="delete"
    inProgress={false}
    onClick={[Function]}
  >
    Delete user
  </Button>
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    fade={true}
    isOpen={false}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    returnFocusAfterClose={true}
    role="dialog"
    scrollable={false}
    toggle={[Function]}
    unmountOnClose={true}
    zIndex={1050}
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      toggle={[Function]}
      wrapTag="div"
    >
      Confirmation
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <p>
        Are you sure you want to 
        <strong>
          delete
        </strong>
         the user?
      </p>
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="secondary"
        onClick={[Function]}
      >
        Cancel
      </Button>
      <Button
        color="primary"
        onClick={[Function]}
      >
        OK
      </Button>
    </ModalFooter>
  </Modal>
</div>
`;

exports[`Component: ConfirmButton ui custom class, color and texts: Component: ConfirmButton => ui => custom color and texts 1`] = `
<div
  className="bigger-better-and-uncut"
  style={
    Object {
      "display": "block",
    }
  }
>
  <Button
    color="info"
    icon="delete"
    inProgress={false}
    onClick={[Function]}
  >
    Delete user
  </Button>
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    fade={true}
    isOpen={false}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    returnFocusAfterClose={true}
    role="dialog"
    scrollable={false}
    toggle={[Function]}
    unmountOnClose={true}
    zIndex={1050}
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      toggle={[Function]}
      wrapTag="div"
    >
      PLEASE SAY YES
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <p>
        Are you sure you want to 
        <strong>
          delete
        </strong>
         the user?
      </p>
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="secondary"
        onClick={[Function]}
      >
        NO
      </Button>
      <Button
        color="primary"
        onClick={[Function]}
      >
        YES
      </Button>
    </ModalFooter>
  </Modal>
</div>
`;

exports[`Component: ConfirmButton ui only button: Component: ConfirmButton => ui => only button 1`] = `
<div
  style={
    Object {
      "display": "block",
    }
  }
>
  <Button
    color="danger"
    inProgress={false}
    onClick={[Function]}
  >
    Delete user
  </Button>
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    fade={true}
    isOpen={false}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    returnFocusAfterClose={true}
    role="dialog"
    scrollable={false}
    toggle={[Function]}
    unmountOnClose={true}
    zIndex={1050}
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      toggle={[Function]}
      wrapTag="div"
    >
      Confirmation
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <p>
        Are you sure you want to 
        <strong>
          delete
        </strong>
         the user?
      </p>
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="secondary"
        onClick={[Function]}
      >
        Cancel
      </Button>
      <Button
        color="primary"
        onClick={[Function]}
      >
        OK
      </Button>
    </ModalFooter>
  </Modal>
</div>
`;

exports[`Component: ConfirmButton ui only icon in progress: Component: ConfirmButton => ui => only icon in progress 1`] = `
<div
  style={
    Object {
      "display": "inline",
    }
  }
>
  <Button
    color="danger"
    icon="delete"
    inProgress={true}
    onClick={[Function]}
  />
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    fade={true}
    isOpen={false}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    returnFocusAfterClose={true}
    role="dialog"
    scrollable={false}
    toggle={[Function]}
    unmountOnClose={true}
    zIndex={1050}
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      toggle={[Function]}
      wrapTag="div"
    >
      Confirmation
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <p>
        Are you sure you want to 
        <strong>
          delete
        </strong>
         the user?
      </p>
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="secondary"
        onClick={[Function]}
      >
        Cancel
      </Button>
      <Button
        color="primary"
        onClick={[Function]}
      >
        OK
      </Button>
    </ModalFooter>
  </Modal>
</div>
`;

exports[`Component: ConfirmButton ui only icon: Component: ConfirmButton => ui => only icon 1`] = `
<div
  style={
    Object {
      "display": "inline",
    }
  }
>
  <Button
    color="danger"
    icon="delete"
    inProgress={false}
    onClick={[Function]}
  />
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransition={
      Object {
        "mountOnEnter": true,
        "timeout": 150,
      }
    }
    centered={false}
    fade={true}
    isOpen={false}
    keyboard={true}
    modalTransition={
      Object {
        "timeout": 300,
      }
    }
    onClosed={[Function]}
    onOpened={[Function]}
    returnFocusAfterClose={true}
    role="dialog"
    scrollable={false}
    toggle={[Function]}
    unmountOnClose={true}
    zIndex={1050}
  >
    <ModalHeader
      charCode={215}
      closeAriaLabel="Close"
      tag="h5"
      toggle={[Function]}
      wrapTag="div"
    >
      Confirmation
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <p>
        Are you sure you want to 
        <strong>
          delete
        </strong>
         the user?
      </p>
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="secondary"
        onClick={[Function]}
      >
        Cancel
      </Button>
      <Button
        color="primary"
        onClick={[Function]}
      >
        OK
      </Button>
    </ModalFooter>
  </Modal>
</div>
`;
